lib/util.ts

```diff
@@ -82,38 +155,86 @@ export class DetailContext implements IContext {
     if (best && best._score > 0) {
       this._propNames.push(...best._propNames);
       this._messages.push(...best._messages);
+      this._failedForks.push(...best._failedForks);
     }
   }
 
   public getError(path: string): VError {
-    const msgParts: string[] = [];
-    for (let i = this._propNames.length - 1; i >= 0; i--) {
-      const p = this._propNames[i];
-      path += (typeof p === "number") ? `[${p}]` : (p ? `.${p}` : "");
-      const m = this._messages[i];
-      if (m) {
-        msgParts.push(`${path} ${m}`);
-      }
-    }
-    return new VError(path, msgParts.join("; "));
+    const fullMessage = flatten(this.getErrorDetails(path).map(errorLines))
+        .join("\n");
+    return new VError(path, fullMessage);
   }
 
-  public getErrorDetail(path: string): IErrorDetail|null {
+  public getErrorDetails(path: string): IErrorDetail[] {
+    let detail: IErrorDetail|null = null;
+    let nested: IErrorDetail;
     const details: IErrorDetail[] = [];
+
+    // As checkers call fail() and return to their parent checkers,
+    // the deepest failures are recorded first.
+    // Go through failures in reverse to start from the root type
     for (let i = this._propNames.length - 1; i >= 0; i--) {
       const p = this._propNames[i];
       path += (typeof p === "number") ? `[${p}]` : (p ? `.${p}` : "");
       const message = this._messages[i];
-      if (message) {
-        details.push({path, message});
+      if (!message) {
+        continue;
+      }
+
+      nested = {path, message}
+      if (detail) {
+        detail.nested = [nested]
+      } else {
+        // This is the root failure, so it will be returned
+        details.push(nested);
       }
+      detail = nested
     }
-    let detail: IErrorDetail|null = null;
-    for (let i = details.length - 1; i >= 0; i--) {
-      if (detail) { details[i].nested = [detail]; }
-      detail = details[i];
+
+    const forkErrors = flatten(this._failedForks.map(fork => fork.getErrorDetails(path)));
+    if (detail && forkErrors.length) {  // don't put an empty array in detail.nested
+      detail.nested = forkErrors;
```

Comment:
Is this safe to do if `detail.nested` was already set (it's set above to `[nested]`)? If yes, could you add a comment, since the details of how errors are collected are a little hard to follow (admittedly were confusing even before forks).